andi/andi.py

```diff
@@ -1,62 +1,302 @@
 # -*- coding: utf-8 -*-
+from collections import OrderedDict
 from typing import (
     Dict, List, Optional, Type, Callable, Union, Container,
-    get_type_hints,
-)
+    get_type_hints, Tuple, cast, MutableMapping)
 
-from andi.typeutils import get_union_args, is_union, get_globalns
+from andi.typeutils import (
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
         else:
-            res[key] = [tp]
+            res[key] = [] if tp is None else [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+PlanList = List[Tuple[Callable, Dict[str, Callable]]]
+
+
+class Plan(PlanList):
```

Comment:
Nice idea of having a Plan as an actual list.

I'm still concerned about inheriting from typing List[Tuple[Callable, Dict[str, Callable]]], it just feels wrong :) In "my world" we should be inheriting from concrete classes and abc classes or mixins, not from types. But I don't know any practical problems with your solution. An alternative is probably inheriting from list, and overriding method(s) with proper signatures, which looks like more code. I don't know an elegant way to handle it.

Reply:
I think is the right thing to do. From the documentation, see this example:
![image](https://user-images.githubusercontent.com/895720/76410957-b618c480-6388-11ea-8a8e-9c71a4fafc5a.png)

So this shows that inheritance from types defined in `typing` is something possible. 

Furthermore, this is the actual implementation of `List`:
```py
class List(list, MutableSequence[T], extra=list):

    __slots__ = ()

    def __new__(cls, *args, **kwds):
        if cls._gorg is List:
            raise TypeError("Type List cannot be instantiated; "
                            "use list() instead")
        return _generic_new(list, cls, *args, **kwds)
```

- It inherits directly from `list`. So it is actually a list. 
- It is also a `MutableSequence[T]` which is from where all annotations for magic methods, etc, are inherited. 

In definitive, `List`, `Dict` are just "Generic variants of builtin collections" as stated in https://www.python.org/dev/peps/pep-0484/#id49

I think it is exactly what we want: a list but with fixed types.

Reply:
Makes sense, thanks for digging in!